lib/Sema/SemaBounds.cpp

```diff
@@ -3285,6 +3288,318 @@ namespace {
+      PrintComparisonSet(B->In, std::string("In"));
+      PrintComparisonSet(B->OutThen, std::string("OutThen"));
+      PrintComparisonSet(B->OutElse, std::string("OutElse"));
+      llvm::errs() << "}\n";
+    }
+  }
+
+  void Reset() {
+    CurrentIndex = 0;
+  }
+
+  void Next() {
+    CurrentIndex++;
+  }
+
+  // This function fills `ComparisonFacts` with pairs (Expr1, Expr2) where
+  // Expr1 < Expr2, Expr1 <= Expr2, Expr2 > Expr1, or Expr2 >= Expr1.
+  // These comparisons correspond to the current block.
+  void GetFacts(std::set<std::pair<Expr *, Expr *>>& ComparisonFacts) {
+    ComparisonFacts = Blocks[CurrentIndex]->In;
+  }
+
+private:
+  ElevatedCFGBlock* GetByCFGBlock(const CFGBlock *B) {
+    return Blocks[BlockIDs[B->getBlockID()]];
+  }
+
+  // Given two sets S1 and S2, the return value is S1 \ S2.
+  ComparisonSet Difference(ComparisonSet& S1, ComparisonSet& S2) {
+   if (S2.size() == 0)
+      return S1;
+    ComparisonSet Result;
+    for (auto E1 : S1)
+      if (S2.find(E1) == S2.end())
+        Result.insert(E1);
+    return Result;
+  }
+
+  // Given two sets S1 and S2, the return value is the union of these sets.
+  ComparisonSet Union(ComparisonSet& S1, ComparisonSet& S2) {
+    if (S1.size() == 0)
+      return S2;
+    if (S2.size() == 0)
+      return S1;
+    ComparisonSet Result(S1);
+    Result.insert(S2.begin(), S2.end());
+    return Result;
+  }
+
+  // Given two sets S1 and S2, this function returns true if the two sets
+  // are not equal.
+  // Equality means that their sizes are the same, and every member of S1 is found in
+  // S2 and vice versa.
+  bool Differ(ComparisonSet& S1, ComparisonSet& S2) {
+    if (S1.size() != S2.size())
+      return true;
+    if (S1.size() == 0 && S2.size() == 0)
+      return false;
+    if (S1.size() > S2.size())
+      for (auto E : S1)
+        if (S2.find(E) == S2.end())
+          return true;
+    if (S1.size() < S2.size())
+      for (auto E : S2)
+        if (S1.find(E) == S1.end())
+          return true;
+    return false;
+  }
+
+  // Given two sets S1 and S2, the return value is the intersection of these sets.
+  ComparisonSet Intersect(ComparisonSet& S1, ComparisonSet& S2) {
+    if (S1.size() == 0)
+      return S1;
+    if (S2.size() == 0)
+      return S2;
+    ComparisonSet Result;
+
+    for (auto E1 : S1)
+      if (S2.find(E1) != S2.end())
+        Result.insert(E1);
+    return Result;
+  }
+
+  // This function returns true if variable V is used in the comparison I.
+  bool ContainsVariable(Comparison& I, const VarDecl *V) {
+    ExprSet Exprs;
+    CollectExpressions(I.first, Exprs);
+    CollectExpressions(I.second, Exprs);
+    for (auto InnerExpr : Exprs)
+      if (const DeclRefExpr *DRE = dyn_cast<DeclRefExpr>(InnerExpr))
+        if (const VarDecl *VD = dyn_cast<VarDecl>(DRE->getDecl()))
+          if (VD == V)
+            return true;
+    return false;
+  }
+
+  // This function looks recursively for comparisons in statement `S`.
```

Comment:
I think the proper invariant is this: this function computes a list of comparisons E1 <= E2 that will be true if the expression S evaluates to true.   Clearly if S is simple direct comparison expression E3 op E4, a comparison can be created if op is one of LE, LT, GE, GT, or EQ.      If S has the form A && B, comparisons can be created for A and B.

Could you add a TODO for handling logical negation the (`!`) operator?

Reply:
Done.